File: Offer 07. 重建二叉树.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

// 递归
class Solution {
    Map<Integer, Integer> map;
    int[] preorder, inorder;

    public TreeNode buildTree(int[] preorder, int[] inorder) {
        if (preorder.length == 0) {
            return null;
        }
        this.preorder = preorder;
        this.inorder = inorder;    
        // 把中序遍历存入哈希表中方便查找
        int n = preorder.length;
        map = new HashMap<Integer, Integer>();
        for (int i = 0; i < n; i++) {
            map.put(inorder[i], i);
        }
        return build(0, n - 1, 0, n - 1);
    }

    public TreeNode build(int preL, int preR, int inL, int inR) {
        if (preL > preR || inL > inR) {
            return null;
        }
        TreeNode root = new TreeNode(preorder[preL]);
        // 查找root在inorder中的位置
        int rootPos = map.get(preorder[preL]);
        // 递归重建树
        // 左子树的长度为：rootPos - inL
        root.left = build(preL + 1, preL + rootPos - inL, inL, rootPos - 1);
        root.right = build(preL + rootPos - inL + 1, preR, rootPos + 1, inR);
        return root;
    }
}

File: Offer 07. 重建二叉树.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> TreeNode:
        mp = dict()
        for i, num in enumerate(inorder):
            mp[num] = i
        
        # root是preorder的角度
        # in_left和in_right都是从inorder的角度
        def build(root, in_left, in_right):
            if in_left > in_right:
                return
            pre_val = preorder[root]
            i = mp[pre_val]
            node = TreeNode(pre_val)

            node.left = build(root + 1, in_left, i - 1)
            node.right = build(root + i - in_left + 1, i + 1, in_right)

            return node

        return build(0, 0, len(inorder) - 1)